### PR TITLE
JENA-1733: SHACL engine, command line validator and Fuseki service.

### PR DESCRIPTION
**Pull request by afs**
It covers Core Constraint Components (sec 4), SPARQL-based Constraints (sec 5) and SPARQL-based Constraint Components (sec 6).

... this does not include shacl-af, shacl-js or shaclc.

It passes the Working Group test suite except for the negative pre-binding tests. It does not reject negative syntax (the full set of syntax rules), though bad shapes may not execute meaningfully.

The W3C WG Test suite is included.

The implementation has
* a Java API
* a command line validator tool, and included in the apache-jena download
* a Fuseki graph validation service

### CODE REVIEW

**Comment by afs** [comment]
Documentation (temporary location for WIP):
https://github.com/afs/morph/blob/master/doc.md

File: jena-shacl/src/main/java/org/apache/jena/shacl/lib/G.java [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+        // Exactly one.
+//        if ( x == null )
+//            throw new RDFDataException("ExactlyOne: None");
+        return x;
+    }
+
+    @SafeVarargs
+    public static <X> X firstNonNull(X ... objects) {
+        int c = 0;
+        X x = null;
```
Comment (kinow) [inline jena-shacl/src/main/java/org/apache/jena/shacl/lib/G.java]: `c` and `x` not used in this method?

File: jena-shacl/src/main/java/org/apache/jena/shacl/lib/Transitive.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        for (; iter1.hasNext();) {
+            Node n1 = iter1.next() ;
+            recurse_1(graph, forward, stepCount + 1, maxStepCount, n1, predicate, visited, output) ;
+        }
+    }
+
+    // A single step of a transitive properties.
+    // Because for SP? or ?PO, no duplicates occur, so works for both strategies.
+    private static Iterator<Node> singleStep(Graph graph, boolean forward, Node node, Node property) {
+        Iter<Node> iter2 = null ;
```
Comment (kinow) [inline jena-shacl/src/main/java/org/apache/jena/shacl/lib/Transitive.java]: Not used `iter2`

File: jena-shacl/src/main/java/org/apache/jena/shacl/validation/ReportEntry.java [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+        Objects.requireNonNull(focusNode,"focus node");
+        //Objects.requireNonNull(message,"message");
+        Objects.requireNonNull(severity,"severity");
+        Objects.requireNonNull(sourceConstraintComponent, "sourceConstraintComponent");
+
+        entry.addProperty(SHACLM.focusNode,       rdfNode.apply(focusNode));
+        entry.addProperty(SHACLM.resultSeverity,  rdfNode.apply(severity.level()));
+        entry.addProperty(SHACLM.sourceConstraintComponent, rdfNode.apply(sourceConstraintComponent));
+
+        if ( messages != null ) {
```
Comment (kinow) [inline jena-shacl/src/main/java/org/apache/jena/shacl/validation/ReportEntry.java]: It's never `null` as it's a private/final `ArrayList` initialized only in this class (other methods use `messages` without defensive check)

File: jena-shacl/src/main/java/org/apache/jena/shacl/lib/ShLib.java [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+                return triple.getObject();
+            case targetSubjectsOf :
+                return triple.getSubject();
+            default :
+        }
+        return null;
+    }
+
+    @SafeVarargs
+    public static <X> X getFirstNonNull(X...xs) {
```
Comment (kinow) [inline jena-shacl/src/main/java/org/apache/jena/shacl/lib/ShLib.java]: Maybe we could use `ObjectUtils.firstNonNull` from Commons Lang (used in jena-arq I think?)

File: jena-shacl/src/main/java/org/apache/jena/shacl/ValidationReport.java [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        report.addProperty(SHACLM.conforms, C.mFALSE);
+        return report;
+    }
+
+    public static ValidationReport fromGraph(Graph graph, Node node) {
+        //RDFDataMgr.write(System.err,  graph, Lang.TTL);
+
+        boolean conforms = G.contains(graph, node, SHACL.conforms, C.TRUE);
+        List<Node> results = G.listSP(graph, node, SHACL.result);
+        if ( conforms && ! results.isEmpty() ) {}
```
Comment (kinow) [inline jena-shacl/src/main/java/org/apache/jena/shacl/ValidationReport.java]: debug/unused code?

File: jena-shacl/src/main/java/org/apache/jena/shacl/engine/constraint/EqualsConstraint.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    public EqualsConstraint(Node property) {
+        super(property);
+    }
+
+    @Override
+    public void validate(ValidationContext vCxt, Shape shape, Node focusNode, Path path,
+                         Set<Node> pathNodes, Set<Node> compareNodes) {
+        for ( Node vn : pathNodes ) {
+            if ( ! compareNodes.contains(vn) ) {
+                String msg = toString()+": not disjoint: vakue node "+displayStr(vn)+" is not in "+compareNodes;
```
Comment (kinow) [inline jena-shacl/src/main/java/org/apache/jena/shacl/engine/constraint/EqualsConstraint.java]: s/vakue/value

File: jena-shacl/testing/std/AFS [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+  
+Changes:
+# Skip
+In: sparql/pre-binding/manifest.ttl
+
+# Optional $shapesGraph not supported.
+<shapesGraph-001.ttl>
+
+## Various pre-binding cases that should be errors
+## but these are unliekly to be unstable laong term.
```
Comment (kinow) [inline jena-shacl/testing/std/AFS]: Is this file like a `TODO` list? If not, if used for something, then maybe s/unliekly/unlikely and s/laong/long ?

File: jena-shacl/src/main/java/org/apache/jena/shacl/engine/ShaclPaths.java [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+
+    // XXX Better error checking.
+    private static Path path(Graph graph, Node node) {
+        if ( node.isURI() && ! C.NIL.equals(node) )
+            return PathFactory.pathLink(node);
+
+        if ( isList(graph, node) ) {
+            List<Node> nodes = G.rdfList(graph, node);
+            if ( nodes.isEmpty() )
+                throw new ShaclParseException("Empty list for path sequnce");
```
Comment (kinow) [inline jena-shacl/src/main/java/org/apache/jena/shacl/engine/ShaclPaths.java]: s/sequnce/sequence

File: jena-shacl/src/main/java/org/apache/jena/shacl/engine/exec/TripleValidator.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+            //case targetClass :
+            case targetNode :
+                return target.getObject();
+            case targetObjectsOf :
+                return triple.getObject();
+            case targetSubjectsOf : {
+                Path path = validationOp.getPath();
+                if ( path != null ) {
+                    if ( !( path instanceof P_Link ) )
+                        throw new ShaclException("Not a predciate path: "+path);
```
Comment (kinow) [inline jena-shacl/src/main/java/org/apache/jena/shacl/engine/exec/TripleValidator.java]: s/predciate/predicate

File: jena-shacl/src/main/java/org/apache/jena/shacl/sys/ValidationGraph.java [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+package org.apache.jena.shacl.sys;
+
+import org.apache.jena.graph.Graph;
+import org.apache.jena.graph.TransactionHandler;
+import org.apache.jena.graph.impl.TransactionHandlerBase;
+import org.apache.jena.shacl.*;
+import org.apache.jena.shacl.validation.ValidationProc;
+import org.apache.jena.sparql.graph.GraphWrapper;
+
+/** A graph that performs SHCAL validation on a graph during transaction "commit".
```
Comment (kinow) [inline jena-shacl/src/main/java/org/apache/jena/shacl/sys/ValidationGraph.java]: s/SHCAL/SHACL ? (this is confusing :eyes: )

**Comment by kinow** [comment]
Tested running in Eclipse `shacl.shacl` from jena-cmds, with `validate --shapes /home/kinow/Development/java/jena/jena/jena-fuseki2/jena-fuseki-main/testing/ShaclValidation/shapes1.ttl --data /home/kinow/Development/java/jena/jena/jena-fuseki2/jena-fuseki-main/testing/ShaclValidation/data1.ttl`.

Output:

```
@prefix sh:    <http://www.w3.org/ns/shacl#> .
@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .

[ a            sh:ValidationReport ;
  sh:conforms  false ;
  sh:result    [ a                             sh:ValidationResult ;
                 sh:focusNode                  <http://example/s1> ;
                 sh:resultMessage              "DatatypeConstraint[xsd:string] : Got datatype http://www.w3.org/2001/XMLSchema#integer : Node 60" ;
                 sh:resultPath                 <http://example/ns#p> ;
                 sh:resultSeverity             sh:Violation ;
                 sh:sourceConstraintComponent  sh:DatatypeConstraintComponent ;
                 sh:sourceShape                _:b0 ;
                 sh:value                      60
               ] ;
  sh:result    [ a                             sh:ValidationResult ;
                 sh:focusNode                  <http://example/s1> ;
                 sh:resultMessage              "maxCount[1]: Invalid cardinality: expected max 1: Got count = 2" ;
                 sh:resultPath                 <http://example/ns#p> ;
                 sh:resultSeverity             sh:Violation ;
                 sh:sourceConstraintComponent  sh:MaxCountConstraintComponent ;
                 sh:sourceShape                _:b0
               ]
] .
```

Removing the block after the comment that says "Invalid", it outputs:

```
@prefix sh:    <http://www.w3.org/ns/shacl#> .
@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix xsd:   <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .

[ a            sh:ValidationReport ;
  sh:conforms  true
] .
```

Had a quick look at the SHACL docs. Unfortunately don't have a use case and don't know SHACL enough, but tried to at least read the code as much as I could before dinner :)

Cheers
Bruno

File: jena-shacl/testing/std/AFS (hunk 7 above)
Comment (afs) [reply jena-shacl/testing/std/AFS]: Thanks for catching that. Renamed as README - this file documents the local differences for unsupported features and not rejecting in negative syntax tests.

File: jena-shacl/src/main/java/org/apache/jena/shacl/sys/ValidationGraph.java (hunk 10 above)
Comment (afs) [reply jena-shacl/src/main/java/org/apache/jena/shacl/sys/ValidationGraph.java]: :-)

**Comment by afs** [comment]
Hi @kinow - thanks for the great review. 

The second commit address all the comments.

The answers to running validation are correct - it's supposed to generate two validation report entries.

**Comment by afs** [comment]
Rebased and aligned to  #595; development commits squashed down to one commit.

**Pull request merged**